### DISCUSSION TITLE
How to make the route helper available in a Vuex store module?

### DISCUSSION THREAD

**Discussion by plweil** [tighten/ziggy#481] [Q&A] [answered]
I'm getting `ReferenceError: route is not defined`.

At the top of my Vuex store module (this is with Vue 2):

```
import Vue from 'vue'
import { ZiggyVue } from 'ziggy'
import { Ziggy } from '@/ziggy'
Vue.use(ZiggyVue, Ziggy)
```

I also have these lines in my app.js file, so I'm not even sure whether I need them in my module or not.

And then I have a bunch of axios calls, e.g.,

```
      const {data} = await axios.get(
        route(
        'api.project.edit_data',
        { id: projectId}
        )
      )
```

**Comment by bakerkretzmar** [comment, marked as answer]
Where are all those axios calls?

**Reply by plweil** [reply]
In actions. E.g.,

```
  async initEditData ({commit, dispatch}, projectId){
    try {
      const {data} = await axios.get(
        route(
        'api.project.edit_data',
        { id: projectId}
        )
      )
    ...
```

To add a little more detail, the action is called in the `created()` hook in `app.js` using a map helper.

```
 async created() {
    if (myObject)
      await this.initEditProject(myObject)
      this.projectLoaded = true
  },
```

`app.js` also has the same Ziggy import lines.

I did see an issue thread (#238) from a while back that talked about using a mixin to use for both components and vuex modules, but the thread seems outdated.

**Reply by bakerkretzmar** [reply]
Okay sorry I haven't used Vuex—where are these "actions"? Like are they all together in one js file somewhere, are they inside individual Vue components, etc.

The Vue plugin basically makes `route()` a method on every component in your entire app, so if you're trying to use the route helper inside the `<script>` tags of a Vue component, you have to call `this.route()`, not just `route()`.

If you're using it in another file you have to import it there, so you would need something like this at the top of the file:

```js
import route, { ZiggyVue } from 'ziggy';
import { Ziggy } from '@/ziggy';

// This isn't required but otherwise you'd have to pass all four arguments every time so that it gets the right config
route = (name, params, absolute, config = Ziggy) => route(name, params, absolute, config);
```

**Reply by plweil** [reply]
That's ok @bakerkretzmar. I'm a little surprised that there aren't more mentions of Vuex in the issues, but ...

My Vuex setup is explained in brief below.

When I import as you just described, I get still get the same error: `ReferenceError: route is not defined.` Apparently

`import route, { ZiggyVue } from 'ziggy'`

needs some tweaking in order to get the route method be to be recognized..

For Vuex, there's a `store.js` file that imports several module files., each of which contains sections for state ("data"), actions ("methods"), and getters ("computed properties").

/resources/components/store
--store.js
/resources/components/store/modules
--app.js
--project.js
etc.

`store.js`

```
import Vue from 'vue'
import Vuex from 'vuex'
import * as project from '@/store/modules/project'
import * as form from '@/store/modules/form'
import * as app from '@/store/modules/app'
import * as notification from '@/store/modules/notification'
import { set, exists, notEmpty, count, spliceFrom, pushTo, makeSet, makeRelation } from '@/utils/VuexUtilities'

import route, { ZiggyVue } from 'ziggy'
import { Ziggy } from '@/ziggy'
Vue.use(ZiggyVue, Ziggy)

Vue.use(Vuex)
const store =  new Vuex.Store({

  modules: {
    project,
    form,
    app,
    notification
  },
})

export default store

```

modules contain the following:
```
export const state = {}

export const actions={}

export const getters={}

```

**Reply by plweil** [reply]
I've tried duplicating and varying the Vuex example offered in issue #238 , but I get this warning:

`WARNING in ./resources/js/route.js 4:9-14
export 'default' (imported as 'route') was not found in 'ziggy' (possible exports: ZiggyVue)`

I'm a little confused by 'route' versus 'ZiggyVue'. It seems as though something ought to work here.

**Reply by bakerkretzmar** [reply]
Ahh my bad I think I might have suggested the wrong imports 🤦🏻 

Try this:

```js
import route from '../...relative-path-to.../vendor/tightenco/ziggy/dist/index.js';
import { ZiggyVue } from '../...relative-path-to.../vendor/tightenco/ziggy/dist/vue.es.js'; // or vue.m.js
```

If that works you might be able to get away with shortening one of those paths using a Mix alias, probably the first one (it looks like you might have already done that).

**Reply by plweil** [reply]
Thanks. well... I think we're getting close :-).  First, I seem to have to use the .es.js suffix for both of these files. Otherwise I get Mix compile errors and warnings. But even with that fixed, I'm getting this error when I load my test page:

`Uncaught ReferenceError: Ziggy is not defined`

I'm not even referencing Ziggy, at least not directly. If we can solve this piece, we might be there.

I also have to add `Vue.use(route, ZiggyVue)`.

**Reply by plweil** [reply]
Thanks for your continued help with this, @bakerkretzmar 

If this is any clue, it seems to be having a problem with Ziggy in `index.es.js`. In the console on page load:

```
index.es.js:1 Uncaught ReferenceError: Ziggy is not defined
    at new rr (index.es.js:1)
    at er (index.es.js:1)
    at Function.Vue.use (vue.esm.js:5132)
    at Object../resources/js/store/store.js (store.js:18)
    at __webpack_require__ (bootstrap:19)
    at Object../resources/js/project_edit.js (interceptors.js:74)
    at __webpack_require__ (bootstrap:19)
    at startup:4
    at Function.__webpack_require__.O (chunk loaded:23)
    at startup:7
```

`store.js`
```
import route from '../../../vendor/tightenco/ziggy/dist/index.es.js'
import { ZiggyVue } from '../../../vendor/tightenco/ziggy/dist/vue.es.js'
Vue.use(route, ZiggyVue)
```
The Mix aliases we talked about are commented out. Without `Vue.use` in this file, it returns "route" is not defined.

The `app.js` file is alsp importing the ZIggy stuff as well as the store.js file:
```
require('./bootstrap')
import Vue from "vue"
import store from '@/store/store'
import { ZiggyVue } from 'ziggy';
import { Ziggy } from './ziggy';
Vue.use(ZiggyVue, Ziggy)
```

**Reply by bakerkretzmar** [reply]
If you're registering the plugin with Vue, the setup you have in `app.js` looks correct. In `store.js` if you need to do the same thing, do it the same way—the first argument to `Vue.use()` should be the plugin and the second argument is a configuration object for the plugin:

```js
import { ZiggyVue } from '../../../vendor/tightenco/ziggy/dist/vue.es.js';
import { Ziggy } from './...path-to-generated-file.../ziggy';
Vue.use(ZiggyVue, Ziggy);
```

But if you're calling `route()` in `store.js` you need to import it and set it up, and it also needs the config object:

```js
import route from '../../../vendor/tightenco/ziggy/dist/index.es.js';
import { Ziggy } from './...path-to-generated-file.../ziggy';

// Not required but sets up a default config so you don't have to pass it every time
route = (name, params, absolute, config = Ziggy) => route(name, params, absolute, config);
```

**Reply by plweil** [reply]
Hey, I think I've got it working now thanks to your last post. I also switched to the npm version (I wasn't aware of it before), which bypasses the need for getting the path right, aliases, etc. The relevant part at the top of my `store.js` file now looks like this:

```
import route from 'ziggy-js'
import { Ziggy }  from '@/routes';
Vue.use(Ziggy)
window.route = (name, params, absolute, config = Ziggy) => route(name, params, absolute, config);
```

Many thanks! This is a really useful package.

**Reply by bakerkretzmar** [reply]
You're welcome! One final note, I'm pretty sure you can remove `Vue.use(Ziggy)` there. If `Ziggy` is the config object with all your routes in it, I don't think that actually does anything since it isn't a Vue plugin.

Glad it's all working for you now!

**Reply by plweil** [reply]
Yes, you're right. It's not needed. I think I was just hedging my bet. Thanks!

**Reply by plweil** [reply]
@bakerkretzmar one quick (last?) question: what is the correct way to import the ZiggyVue plugin for components when using the npm package? This is giving me an error:

```
Uncaught TypeError: Cannot read properties of undefined (reading 'install')
    at Function.Vue.use (vue.esm.js:5129)
    at Module../resources/js/dashboard_admin.js (dashboard_admin.js:16)
    at __webpack_require__ (bootstrap:19)
    at startup:4
    at Function.__webpack_require__.O (chunk loaded:23)
    at startup:5
    at startup:5
```

`dashboard_admin.js` (root Vue file)

```
import { ZiggyVue } from 'ziggy-js'
import { Ziggy } from '@/ziggy'
Vue.use(ZiggyVue, Ziggy)
```

**Reply by bakerkretzmar** [reply]
I think you need `import { ZiggyVue } from 'ziggy-js/vue'`, if that doesn't work try the whole filename, `ziggy-js/vue.es.js`.

**Reply by plweil** [reply]
`import { ZiggyVue } from 'ziggy-js/vue'`

Perfect. Thanks again!
